Data Analysis: Do high emissions predict reporting non-compliance?

colton-lapp opened the pull request:
# Description

This pull request creates:
1. A jupyter notebook file with data analysis findings
2. A blog post describing the findings in the jupyter notebook file 

The blog post investigates a question raised in #114 - does poor performance correlate with non-reporting? The short answer is no, I didn't find that pattern in the data.

The data analysis in the Jupter notebook consists of the following steps:
1. Create some basic data viz showing variables of interest and compliance trends over time
3. Create lagged variables of emissions last year and the emission trends from 2 years ago to 1 year ago
4. Create graphs comparing mean/median GHG intensity last year and GHG trend from 2 years ago to 1 year ago vs reporting compliance, showing basically no difference
5. Run a regression with a single control variable (square footage) to confirm there is no statistically significant relationship
6. Run some robustness checks by dropping outliers and dropping covid and repeating steps 3-4; still no significant results

These findings are then summarized in a new blog post. 

A couple other notes:
- I used the graphing package plotly to make html graphs that are interactive, and embedded those in my blog. Because the html graphs allow you to over over individual data points and display info, they take up a decent amount of space (between 1-20 mb). This also makes the Jupyter notebook file larger, but I tried to cut down the size by making some of the plots static image files
- I don't know anything about javascript or html so I relied on Gen AI to do a lot of the coding for embedding and rendering interactive html graphs and fetching regression results from a JSON file. This could probably use some serious attention
- I've introduced some new dependencies for data visualization and am not sure how this is managed in the project

This is my *first time creating a PR* for a public repo and for this project specifically so happy to restructure any work or accept any feedback! I'm expecting some heavy feedback on files committed (i.e. new packages used in requirements.txt, python virtual environment, directory structure). 

Fixes #114 

# Testing Instructions

I would recommend pulling, running docker-compose up and looking at my blog. Additionally, check out the Jupyter notebook to verify that I'm analyzing the correct variables and don't have data analysis mistakes, etc. **To see the interactive html graphs in the notebook, you have to view the Jupyter file in NBViewer as it won't render in Github**

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

vkoves commented:
@colton-lapp - I meant to comment when I pushed up my fixes - I've added some date stamps to the blog posts and reodered it so yours comes first (since it's newer). I'm fine with that file size, but it looks like there's some responsiveness issues with the graphs - if you can fix those, I'm good with it, but otherwise we could move to images. Here's an example:

| Desktop (shows scrollbar) | Mobile (cut-off) |
| --- | --- |
| ![Screenshot from 2025-01-20 16-37-50](https://github.com/user-attachments/assets/1434ad84-1fb0-4f1a-aae0-96c8ac291a3d) | ![Screenshot from 2025-01-20 16-38-02](https://github.com/user-attachments/assets/bb3a1a26-cdfa-4e4c-93eb-0e2d865983fe) |

Also is there a way to note dependencies for you Jupyter notebook? I tried running it locally but had to manually install dependencies like `plotly`, which aren't in our `requirements.txt`. Maybe you can add some instructions at the start of your notebook and maybe a `requirements.txt` file? I don't know what's typical there

vkoves commented:
@colton-lapp - wanted to check in on this, it looks like Pytest is failing, and if you need help using the graph images just on mobile, let me know!

I also showed the preview to one of our partner's at Climate Reality Chicago (who asked for this research question), and they were a big fan, and said:

> I love, love, love the visuals and the way you methodically walk the reader through your line of thinking! And I can imagine the “Distribution of GHG Intensities” image being useful for us in other contexts in the future.

> The only change I might suggest is that you have 2 graphs immediately above “Results: No noticeable difference between groups” that mention “last year” and “this year” in their titles. I**t might be good to replace those with numerical years so that people don’t have to check the blog post publication year / can refer to those graphs in other contexts** without creating ambiguity about which year’s data are included. Just a thought.

I agree with her feedback on adding years, and I think that would be a simple improvement to each graph, just to make sure if it gets screenshotted and shared it's really clear what data we used. Even adding Chicago in there might be helpful, but 🤷🏻‍♂️

vkoves commented on src/pages/blog/GHGIntensityPredictCompliance.vue:
```diff
@@ -0,0 +1,430 @@
+          rel="noopener noreferrer"
+          >don't report emissions data.</a
+        >
+        Is there a pattern to which buildings fail to report? Anecdotally, it's
```
I'd change this "Anecdotally..." sentence to "Our team has noticed that some high emissions buildings stop reporting, while more efficient buildings tend to keep reporting year after year." There unfortunately has not been real press coverage of this data 😭

colton-lapp commented:
@vkoves 

Finally got some time to work on this and addressed everything you mentioned I think:

- [ ] Tests passing with new requirements.txt file (note, you wanted a seperate requirements.txt file but the tests seem to use src/data/requirements.txt, so I had to change that one. I have a seperate file in src/data/analysis/predict_compliance_requirements.txt that is identical which I was hoping to use. I'll let you sort this out in the tests if you want
- [ ] Using basic javascript to check for mobile, and showing static PNG files if it's mobile
- [ ] Adopted your phrasing changes. Linked to the dataset when I mention it the first time
- [ ] Added authorship to the top (feel free to restyle) which links to my Github (I don't really have social media, happy to take the link off)
- [ ] I changed the graphs as suggested to reference specific years. Note - in the previous graphs, I was actually showing the data for all the years, hence the vague language (i.e. saying "emissions last year for compliance this year had data points with emissions in 2018, compliance in 2019 AND emissions in 2019, compliance in 2020, etc). I subset the dataframe to only show the most recent year to make it clearer and updated the title
- [ ] Some other small tweaks to graphs and stuff

vkoves reviewed (changes requested):
@colton-lapp - the resize change works great, but if you test your preview on mobile, the font sizes are too small to read. Can you re-export those PNGs with larger font sizes targeted to mobile? You could also move the legend below the graphs (if that's an option) and you need more room.

![Screenshot from 2025-02-10 20-47-06](https://github.com/user-attachments/assets/0dbe351e-4b70-428b-b944-642d9d52983d)

I've also made a few minor tweaks to the blog post to improve the styling, particularly the regression table:

![Screenshot from 2025-02-10 21-51-02](https://github.com/user-attachments/assets/acf864bb-3d2e-4c87-898d-0dd94650993e)

vkoves commented on src/pages/About.vue:
```diff
@@ -25,19 +25,20 @@ export default class About extends Vue {
-      <p class="constrained">
-        Electrify Chicago is an independent open-source project based out of
-        <a href="https://chihacknight.org" target="_blank"
-          >Chi Hack Night <NewTabIcon
```
@colton-lapp - did you mean to remove this Chi Hack Night link, or was it a faulty merge?

vkoves reviewed (changes requested):
A few minor comments - I think two little things on my end I want to tweak the blue background and border radius of the caption blocks, and make sure to add your new article to the homepage!

I also see you've tweaked the graph images, but have you tried them on mobile with the [Netlify preview](https://deploy-preview-147--radiant-cucurucho-d09bae.netlify.app/blog/ghg-intensity-predict-compliance/)? The text still seems a little small to me, but maybe there's nothing more we can do there, and folks will have to flip their phone or zoom in 🤷🏻‍♂️ (graphs are hard on mobile no matter what)

colton-lapp commented on src/pages/blog/GHGIntensityPredictCompliance.vue:
```diff
@@ -322,81 +339,103 @@ export default class MillionsInMissedFine extends Vue {
+        but most have less than a dozen buildings in the data. To see this graph
+        with all the different building types, please view the
+        <a
+          href="https://nbviewer.org/github/vkoves/electrify-chicago/blob/compliance-analysis/src/data/analysis/GHG_intensity_compliance_correlation.ipynb"
```
Yeah it's the same link. Sorry I needed to clean this up. At first I just said "refer below" but that seemed annoying so I added the link again. I'll choose one or the other.

vkoves replied:
I think having the link there is great, so I'd just change it to not say look below

colton-lapp commented on src/pages/blog/GHGIntensityPredictCompliance.vue:
```diff
@@ -473,20 +525,47 @@ export default class MillionsInMissedFine extends Vue {
+          At Electrify Chicago, we’ve noticed that few politicians and
+          policymakers are even aware of the existence of the Chicago Building
+          Benchmarking Data. So, it’s not surprising that we see no pattern—why
+          hide emissions if no one is paying attention?
```
A rhetorical question, yes. "It's not surprising there's no pattern - after all, why would buildings hide emissions if no one is paying attention?". I'll reword it.

colton-lapp commented on src/pages/blog/GHGIntensityPredictCompliance.vue:
```diff
@@ -543,11 +622,24 @@ export default class MillionsInMissedFine extends Vue {
   }
 
+  .graph-caption {
+    background-color: rgba(187, 212, 255, 0.485);
```
Apologies - still getting the hang of how to properly set / store css variables. Should I use a color from the electrify chicago global palette? Or introducing a new color is fine?

vkoves replied:
You can add a new one for this file, if you'd like! Just `$lightBlue: #hexcode` should do it. You can use Chrome DevTools to mess with colors and flip it from rgb to hex.

vkoves commented:
@colton-lapp - I've added your article to the homepage on this branch and made a few minor fixes to your blue panels, including adding a new `$blue-light` color that they use. I think I'm happy with this once you make those language fixes, and I think we can skip more mobile graph tweaks unless that's really easy to do!

**The one last todo:**

- **Make sure to update the publish date** - this is hard-coded on the homepage, blog home, and article page, so make sure we update it when it's about to get merged. We could more elegantly handle this with a Gridsome collection or a constant, but I'm punting that 😆

colton-lapp commented:
@vkoves Thank you for fixing the styling! I'm hoping I can hop on to the call late tonight and get this finished, but if not, I will in the next day or two.

colton-lapp commented:
Hey @vkoves, I've made the requested changes (as well as a few others) and this PR might be ready to be merged. Specifically, I did the following:

- [ ] Reworded all sentences you commented on in PR comments
- [ ] Reworded some additional sentences I thought were off
- [ ] Updated publish date to **March 1st** in blog, blog index page and home page
- [ ] Updated styling of author line to put it below publish date and reduce bottom margin 
- [ ] Added one more significant digit to regression coefficients to avoid 0.000 for some coefficients

You can look at the code changes for specific rewordings. Feel free to merge if things look good to you or continue to edit styles/wording in any way that you see fit.

vkoves commented:
@colton-lapp - this looks great! The only last thing is can you sort out the `requirements.txt`? I think Sasha auto-generated his somehow, so I'm not sure how to merge those

colton-lapp commented:
@vkoves I spent way too much time trying to fix this haha. I was running into issues locally trying to sort out dependency management (requirements.txt). I'm no expert here and Python is notoriously really annoying for this stuff. I think there may have been issues for me locally because we don't have our virtual environments and requirements.txt files in our root dir. Not positive that this is what was giving me issues, but in most repositories I see for Python projects you also see these things at the root. So I did the following:

- [ ] I moved all Python virtual environment stuff to the root directory
- [ ] I created a Markdown file called `PythonDev.md` that explains how to set up virtual python environments for new users
- [ ] I migrated the virtual environment manager to a modern tool called `uv` that they use at my work and I think is highly regarded. 
- [ ] I included instructions on how to make two kinds of virtual environments - one with core dependencies (more lightweight) and one with additional dependencies for data analysis. This isn't natively supported super easily in uv but it's not too hard and I have it working on my local machine (instructions in PythonDev.md)
- [ ] I created a pyproject.toml file to record these different dependency sets. pyproject.toml is a more modern way to gather requirements for python projects
- [ ] You can still create the necessary requirements.txt file FROM the pyproject.toml/virtual environments. Instructions in PythonDev.md
- [ ] I made all the Python setup instructions reference Python 3.13 which our automatic test deploys use. Not sure if that was an intentional choice. Maybe we should add a .pythonversion file at the root to indicate this

If you're okay with this, I think we can just delete all previous requirements.txt files in main and override it with my changes. Maybe someone who knows more about Python dependency control could review these changes.

vkoves commented:
@colton-lapp - that approach seems sound to me, thank you for all of your work sorting that out. I'm no Python expert, so I'm deferring to you here 😆 

Once you fix the merge conflict, I'll be happy with these changes. Also should we link your new markdown file in the `README`?

colton-lapp commented:
@vkoves I think something got messed up on my branch with a failed merge from main into compliance-analysis like 20 commits ago... When I try to do git pull origin main --rebase, there are so many conflicts. I can try to debug this later but it seems like a big mess (I'm also kind of a version control noob). I don't even have src/data/requirements.txt in my local branch anymore and git won't let me commit that deletion... not sure what's happening there. 

I don't have time to debug this today. If you know of a quick fix, feel free to do whatever is best. 

I also updated run_all.sh to use a Python virtual environment (defaulting to electrify-chicago/.venv). Not sure how dependencies were working before that? But now it should activate a virtual environment with core dependencies when running data processing scripts. I also updated README.md to mention python dependencies as well as in the src/data/README. 

Sorry I've made a mess with the version control... tbh I have no idea what's going on.

vkoves commented:
@colton-lapp - before dependencies just worked via a `pip install` through the dockerfile, take a look at the `README.md`. Let me see if I can sort out these conflicts.

vkoves commented:
@colton-lapp - just merged, but you modified `run_all.sh`, which no longer exists. I migrated to using Python to make things cleaner, via  `run_all.py`. I apologize for the chaos 😭

colton-lapp commented:
@vkoves Oh no I was trying just now to merge into main haha I shouldn't have messaged and kept working. Was just trying to fix it. 

Feel free to continue fixing things. My new method of using uv for python dependencies can still generate requirements.txt for backwards compatibility with Pytest I think but I can look into this later. You can freeze depedencies from pyproject.toml into a requirements file (in my PythonDev.md file). Okay I'll stop working on this for real now

vkoves commented:
@colton-lapp - I'm honestly not sure I know how best to fix it, so do you have an approach that could work, or would you like me to try? I just hit merge and did some best guesses to try and help, but you can feel free to revert that and force push 😆

colton-lapp commented:
@vkoves got the tests to pass! Not sure why we were using Python 3.9 before in tests (no idea what a good strategy is for choosing a python version for testing) but I updated PythonDev.md and pytest.yml to all use Python 3.12, which is one version away from the most recent version Python 3.13 that's only a couple months old. Looking at the code changes, I think everything is okay with the merge into main but I can look at it a little more closely later today maybe.

vkoves commented:
@colton-lapp let's take a look tomorrow during our meeting! My only safety check I'll want to run is that I can still boot up the platform with docker, so we should be good to go if you tested that 🤞🏻

vkoves commented:
@colton-lapp - make sure you test running `pytest` via Docker using the `README` commands _after_ updating the Docker Python version to 3.12 (see line `Dockerfile:2`).

vkoves commented on .github/workflows/pytest.yml:
```diff
@@ -18,12 +18,17 @@ jobs:
-          cd src/data
           python -m pip install --upgrade pip
-          pip install -r requirements.txt
+          pip install uv pytest
```
I think you'll need to copy these to the `Dockerfile` so it uses `uv`.

vkoves replied:
This might be helpful, but maybe not 😆 

https://docs.astral.sh/uv/guides/integration/docker/#installing-uv

vkoves replied:
**Update:** Can we switch the GitHub action to not use `uv`, and just use `pip install .` so that it's faster?

colton-lapp commented:
@vkoves Okay I think everything is good to go! I updated the Readme and the Python dev guide. The Python dev guide is more verbose now explaining when to use a Python virtual environment instead of a Docker container. I also updated one of the commands in the Readme under **Run Data Processing**, the command was indicating the function was a bash script, not a python script, so I changed it to `docker-compose run --rm electrify-chicago python3 run_all.py`. I have all the README commands working now, and also am able to run my notebook using my Python Developer guide to install dependencies. Finally, the pytest file has been simplified to just use pip. 

If you pull the project and everything runs on your machine too, we should be good to go! Just need to update the publishing date. I'll (ambitiously) set that for tomorrow, but feel free to change it if you don't get to it by then.

colton-lapp commented:
@vkoves Found one more place we need to sync Python versions and dependencies - in the Netlify preview. I added a netlify.toml to specify build instructions, including that it should use Python 3.12 and also source dependencies from pyproject.toml

vkoves commented:
@colton-lapp - huh, Netlify shouldn't need Python at all, it's just doing a `yarn build`. We only use Python for notebooks and data processing, so that's not relevant to the build process.

vkoves reviewed (approved):
Looks great now, thank you for all your hard work! I tried to make Netlify install Python dependencies, but apparently it's just _not possible_ - if it finds a `requirements.txt`, it installs things, I just didn't realize this before 😆 🤦🏻‍♂️
